### DISCUSSION TITLE
Must Support - что это значит

### DISCUSSION THREAD

**Discussion by shamilnn** [fhir-ru/core#38] [General]
Некоторые элементы в профайлах могут/будут отмечены как Must Support (красный квадратик с буковкой “S”). Где-то в Ru Core IG должно быть описание какой смысл в это вкладывается. Так, например:

- FHIR R4 даёт весьма расплывчатое описание что бы это могло значить - https://hl7.org/fhir/R4/conformance-rules.html#mustSupport

- US Core более специфичен - https://www.hl7.org/fhir/us/core/conformance-expectations.html#must-support-elements

- Canadian Baseline ещё более специфичен - https://build.fhir.org/ig/HL7-Canada/ca-baseline/branches/master/general-guidance.html#cardinality-and-mustsupport-definitions 

ВОПРОС: Нужно ли отмечать как Must Support обязательный (т.е. cardinality 1..1, 1..*) элемент и что это будет означать для отправляющей и получающей стороны.

**Comment by Vrodvi** [comment]
Расплывчатое описание обусловлено гибкостью стандарта.
В пояснении как раз сказано, что конкретное значение этой отметки ДОЛЖНО быть чётко обозначено в профиле. 
> When a profile does this, it SHALL also make clear exactly what kind of "support" is required, as this could involve expectations around what a system must store, display, allow data capture of, include in decision logic, pass on to other data consumers, etc.

На мой взгляд таблица из канадской спецификации с перебором возможных комбинаций кардинальности и отметки "Необходимо поддерживать" наиболее наглядна, предлагаю создать аналогичную таблицу и для Ru Core.

> ВОПРОС: Нужно ли отмечать как Must Support обязательный (т.е. cardinality 1..1, 1..*) элемент и что это будет означать для отправляющей и получающей стороны.

Автоматически отмечать элементы как Must Support на основании кардинальности **нельзя**, это специально сделано отдельной настройкой.
Кардинальность говорит лишь о необходимости наличия элемента в структуре, но в элементе может быть вставлено фиксированное значение или "заглушка", которую можно игнорировать и не использовать при анализе переданных данных.

Также с помощью атрибута Must Support можно обозначить в спецификации ситуацию, когда элемент ДОЛЖЕН быть заполнен данными, если они имеются, а при отсутствии данных элемент не будет присутствовать. Эту ситуацию не обозначить только с помощью кардинальности, т.к. насильственно объявленная обязательность 1..1 в таком случае приведёт к заполнению элемента фиктивными данными, которые реально нельзя будет использовать.
Кроме того в случае "открытых" профилей ограничений не возможно будет обозначить какие опциональные элементы (0..1 или 0..*) остались из базовых требований FHIR, а какие опциональные элементы определены профилем как необходимые для использования - их как раз можно обозначать через атрибут Must Support.

Ранее в стандарте CDA была доступна ещё более тонкая настройка структуры элементов, которая позволяла обозначить не только обязательность наличия элемента в структуре, но требования по заполнению элемента КОРРЕКТНЫМИ данными, так была возможность указать
R - элемент должен быть заполнен данными, а если данных для элемента нет, то в нём нужно указать nullFlavor
M - элемент обязан быть заполнен данными, а если данных нет, то такой элемент нельзя заполнять заглушкой\фиктивными данными, т.е. невозможно создать экземпляр данных в соответствии с требованиями

![image](https://user-images.githubusercontent.com/92928137/143400943-b9553717-a292-449d-9740-d45bddac03b4.png)

В FHIR данная детальная настройка утеряна, поэтому становится особенно критично определение правил описания структуры с использованием тех инструментов, что остались, а именно - кардинальности и отметки Must Support.

**Comment by Vrodvi** [comment]
Уточнение после совещания от 30.11.2021:
было высказано, что каждый профиль должен быть снабжён как минимум двумя примерами ресурсов - минимально заполненный и максимально заполненный.

С минимально заполненным примером скорее всего сложностей не будет.

Но для формирования максимального примера уже важно учитывать отметки MUST SUPPORT - т.к. профили будут открытые, то все опционально определённые атрибуты (0..* 0..1 и пр.) ресурса будут доступны для заполнения данными, но не все из них являются частью профиля.

Некоторые из опциональных атрибутов ресурса должны быть помечены как MUST SUPPORT  - и именно такие атрибуты должны быть заполнены в "максимальном примере", в то время как остальные опциональные атрибуты могут отсутствовать даже в "максимальном примере".

В связи с этим прошу всё же пересмотреть отношение к отметке "MUST SUPPORT" как к атавизму.
Или предложить вариант обозначения опциальных атрибутов профиля без этой отметки.

**Reply by eugenekogan-spb** [reply]
@Vrodvi,  для описанной вами цели вроде бы хватает кардинальности, Must Support = (0..1). Вот пример. Пусть в профиле Ru Core будет определен ресурс Organization с полями Полное наименование (1..1), Краткое наименование (0..1), Цвет огнетушителей (0..1). В профиле для ФРМО все эти атрибуты будут (1..1). А вот в IG для записи на прием к врачу, который мы когда-нибудь сделаем, надо будет указать, что краткое наименование может быть поддержано (0..1), а цвет огнетушителя не имеет отношения к делу (0..0).

**Reply by Vrodvi** [reply]
ФРМО - это уже следующий уровень профилей, не RU CORE.
В RU CORE большинство полей должно быть опционально, чтобы в руководствах следующего уровня уже их ограничивать.

На конкретном примере:

вот базовая спецификация ресурса "Организация":
```
.. Organization	
... identifier	ΣI	0..*	
... active	?!Σ	0..1
... type	Σ	0..*	
... name	ΣI	0..1
... alias		0..*	
... telecom	I	0..*	
... address	I	0..*	
... partOf	Σ	0..1	
... contact		0..*	
.... purpose		0..1	
.... name		0..1	
.... telecom		0..*	
.... address		0..1	
... endpoint		0..*	
```

Все поля опциональные, но не все эти поля будут описаны в спецификации RU CORE EXAMPLE,
Предположим, что в RU CORE EXAMPLE определены следующие требования как Вы указали:
```
Полное наименование (1..1) (name)
Краткое наименование (0..1) (alias)
```

Все остальные поля в RU CORE EXAMPLE нельзя определять как запрещённые (0..0), т.к. базовая спецификация должна быть максимально общая и открытая!


Получается, что в соответствии с установленными в RU CORE EXAMPLE требованиями 
пример минимально заполненной структуры - это
.. Organization	
... "name": "пример полного наименования МО"

, а пример максимально заполненной структуры - это 
.. Organization	
... "name": "пример полного наименования МО"
... "alias": "пример короткого наименования №1 МО"

В этих примерах мин. и макс. структуры не будет остальных опциональных полей (type, telecom, и прочие), т.к. они спецификацией RU CORE EXAMPLE напрямую не профилируются (открытый профиль).

Получается, что просто кардинальности не достаточно.

Для обозначения требований этого профиля должно быть указано
```
.. Organization	
... name		1..1	MUST SUPPORT
... alias		0..1	MUST SUPPORT


```

**Reply by eugenekogan-spb** [reply]
Похоже, все это означает, что наше решение загнать все поля в базовый профиль не идеальное, оно противоречит объектному подходу. По правилам наследования, в атрибуты наследников надо вводить в наследниках, а не просто ставить им обязательность.
В нашем случае на базовом уровне нужна организация как то, что оказывает медицинскую помощь. Именно в этой роли она участвует в процессах записи на прием, учета оказанной медпомощи и пр. 
"Цвет огнетушителя" свойственен организации с точки зрения руководителя или собственника бизнеса. Ему надо знать все ресурсы этой организации, плановые и фактические объемы помощи и точки расстановки огнетушителей. То есть, по правилам наследования, не надо это включать в базовый уровень.

**Comment by eugenekogan-spb** [comment]
@niquola а как указать разницу между ситуацией, когда при отсутствии информации надо указать nullFlavor, и ситуацией, когда при отсутствии информации нельзя передавать элемент.

**Reply by niquola** [reply]
Но это точно не must support

**Reply by eugenekogan-spb** [reply]
Ок. Тогда я вижу следующие способы:
1) описываем эти детали в IG текстом
2) делаем какой-то значок в Structure Definition для обозначения этой детали
3) не обращаем внимание на такие различия

**Reply by shamilnn** [reply]
Это может зависеть от типа данных элемента и cardinality.

Например, если элемент обязательный (1..1, 1..*), тип данных _CodeableConcept_ и binding [_Extensible_](http://hl7.org/fhir/R4/terminologies.html#extensible) или выше (Required), то элемент ДОЛЖЕН быть заполнен в соответствии с предлагаемым ValueSet. Если, при отсутствии информации можно передавать NullFlavor, то необходимо включить код(ы) NullFlavor в ValueSet подобного элемента.

Если информации нет, и элемент опциональный (0..1, 0..*), то FHIR запрещает передавать пустой элемент (ситуация, когда при отсутствии информации нельзя передавать элемент).

**Comment by eugenekogan-spb** [comment]
По решению, принятому [на встрече 6 сентября](http://fhir.ru/meetings.2022-09-06), возвращаемся к вопросу о must support.

1. Следует включить [в глоссарий RuCore](http://fhir.ru/terms) понятия, описывающие требования к обязательности информации. Примером такого описания являются материалы, приведенные выше в этой дискуссии.
@shamilnn: примером являются также [материалы Canadian Baseline](http://build.fhir.org/ig/HL7-Canada/ca-baseline/branches/master/general-guidance.html#cardinality-and-mustsupport-definitions) (там среди прочего отличные рассуждения о том, где вводить must support - на уровне ядра или наследованных руководств по реализации - прим.ред.)

2. После этого следует включить в схемы профилей, где это требуется, флаг must support.

**Reply by Vrodvi** [reply]
@eugenekogan-spb @shamilnn 
https://build.fhir.org/ig/HL7-Canada/ca-baseline/
на данный момент недоступен, просьба проверить в чём дело
![image](https://user-images.githubusercontent.com/92928137/190100189-4440a99e-34ba-445f-b1d6-d15ddb42a7ef.png)

**Comment by eugenekogan-spb** [comment]
Одно из предложений [обсуждения 13.09.22](http://fhir.ru/meetings.2022-09-13), конкретно для руководства по реализации передачи СЭМД (@Vrodvi):
1) там, где СЭМД требует наличие значения, ставим кардинальность 1.., запрещаем расширения (то есть, не удастся передать NullFlavor); 
2) там, где СЭМД требует при отсутствии значения передавать атрибут и указывать NullFlavor, ставим кардинальность 1.., добавляем в атрибуте NullFlavor Extension, требуем его заполнения.
3) там, где СЭМД позволяет при отсутствии значения не передавать атрибут, ставим кардинальность 0.., не передаем атрибут;
Надо проверить, что в варианте 2 пройдет валидация.

**Reply by Vrodvi** [reply]
Если быть точнее, то предложение было в том, чтобы для руководств по реализации СЭМД'ов при их переложении в формат FHIR описать типы требований к элементам и дать по каждому варианту подходящий способ представления этих требований средствами профилей FHIR.

Вот таблица типов требований из СЭМД (данные типы являются общими для всех видов СЭМД'ов) с описанием возможного представления средствами FHIR:

| **Короткое требование** | **Полное требование**                                                                                                                                                                                                                                                                                                                 | **Кардинальность** | **Отметка MustSupport (S)<br>** | **Возможность указать расширения с причиной отсутствия данных<br>(<br>http://hl7.org/fhir/extension-iso21090-nullflavor.html<br>или<br>http://hl7.org/fhir/extension-data-absent-reason.html<br>)<br>** | **Примеры заполнения, когда данные имеются<br>(указать для случаев с простыми и кодированными данными)** | **Примеры заполнения, когда данные отсутствуют<br>(указать для случаев с простыми и кодированными данными)** |
|-------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------|---------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------|
| "ОБЯЗАН"/"НЕ ОБЯЗАН"    | обязательное требование включения/исключения<br>элемента. Нарушение приведет к несоответствию сформированного CDA документа<br>требованиям данного руководства. Ключевое слово "ОБЯЗАН"/"НЕ ОБЯЗАН"<br>используется для описания варианта, когда элемент не может иметь пустое наполнение<br>и НЕ может содержать атрибут nullFlavor. | 1..                | S                               | ЗАПРЕЩЕНО                                                                                                                                                                                               |                                                                                                          |                                                                                                              |
| "ДОЛЖЕН"/"НЕ ДОЛЖЕН"    | обязательное требование включения/исключения<br>элемента. Нарушение приведет к несоответствию сформированного CDA документа<br>требованиям данного руководства. Ключевое слово "ДОЛЖЕН"/"НЕ ДОЛЖЕН"<br>используется для описания варианта, когда элемент не может иметь пустое наполнение,<br>но может содержать атрибут nullFlavor.  | 1..                | S                               | разрешено                                                                                                                                                                                               |                                                                                                          |                                                                                                              |
| "СЛЕДУЕТ"/"НЕ СЛЕДУЕТ"  | рекомендованные требования для наилучшего<br>соответствия данному руководству ("best practice"). Нарушение не приведет к утрате<br>соответствия, однако разработчик должен иметь веские причины, чтобы<br>включить/исключить элемент.                                                                                                 | 0..                | S                               | разрешено                                                                                                                                                                                               |                                                                                                          |                                                                                                              |
| "МОЖЕТ"/"НЕ МОЖЕТ"      | необязательное требование. Элементы могут быть<br>включены или опущены на усмотрение разработчика реализации без каких-либо<br>последствий.                                                                                                                                                                                           | 0..                |                                 | разрешено                                                                                                                                                                                               |                                                                                                          |                                                                                                              |

Колонки с примерами оставил пустыми - на их подготовку нужно время.

Плюс потребуются дополнительные пояснения по конкретному использованию расширения отсутствия данных, т.к. для простых и кодированных данных подход разный, подробнее можно посмотреть как эти нюансы описываются в US Core -
https://build.fhir.org/ig/HL7/US-Core/must-support.html
https://build.fhir.org/ig/HL7/US-Core/general-requirements.html#missing-data

**Reply by OlegPenzin** [reply]
Коллеги, @eugenekogan-spb ,в рамках клинической рабочей группы еще раз обсудили вопрос - с учетом большой работы по воспроизведению СЭМДов на FHIR - мы считаем оправданным следовать описанной выше логике с увязкой кардинальности и MustSupport ключевым словам руководств

**Reply by eugenekogan-spb** [reply]
Виталий, @Vrodvi Олег, @OlegPenzin, я бы предложил указать MustSupport = S для последней строки. У конкретного экземпляра ресурса атрибут может быть опущен, но в целом передающая и принимающая система обязаны уметь обрабатывать такой атрибут.

**Reply by OlegPenzin** [reply]
@eugenekogan-spb я бы предпочел не проставлять MustSuppoort для ключевого слова МОЖЕТ - в текущем варианте каждому ключевому слову соответствует своя уникальная комбинация Кардинальность * MustSupport * nullFlavor

также  МОЖЕТ - это уже совсем опциональные атрибуты

**Reply by eugenekogan-spb** [reply]
Я вижу два аргумента в пользу поставить:
1) Если не поставить MustSupport, мы теряем информацию о том, что данный атрибут включен в состав СЭМД. 
2) Поскольку атрибут МОЖЕТ быть передан, то принимающая сторона обязана уметь принимать его, что и выражается атрибутом MustSupport. 

Похоже, тонкое различие "может-следует" должно выражаться не нашими флагами, а комментарием о соответствии лучшим практикам. Примерно так же, как иные рекомендации по обеспечению качественной информации.

**Comment by Vrodvi** [comment]
@eugenekogan-spb Евгений, вы просили фрагмент документации FHIR, из которого следует что значение атрибутов не может быть пустым:
http://hl7.org/fhir/json.html 
![image](https://user-images.githubusercontent.com/92928137/194066305-79629096-019d-445d-8854-bb98ed1e7f5a.png)

**Reply by eugenekogan-spb** [reply]
@Vrodvi Виталий, спасибо!

**Comment by eugenekogan-spb** [comment]
Комментарий "после драки".
Никакое решение по данной дискуссии принято не было. Однако при разработке профилей для СЭМД протокола лабораторного исследования в тоге лабораторная группа решила указывать флаг MustSupport во всех атрибутах, которые поименованы в руководстве по заполнению СЭМД, без какой-либо связи с требованиями обязательности заполнения информации. Логика была следующая: если атрибут поименован в руководстве, то любая система вправе такое поле передать. Значит, все системы обязаны уметь принять и обработать такое поле.
